### PR TITLE
Win arm64

### PR DESCRIPTION
**Pull request by rahku**
Arm64 SDK is still to be released. Using a private drop of the toolset and these changes I am able to successfully build arm64 builds for windows. We need to set certain environment variables to override the location of toolset and sdk.

### CODE REVIEW

**Comment by rahku** [comment]
Please review.
@gkhanna79 @jkotas @janvorli @briansull

**Comment by jkotas** [comment]
Could you please rebase away the merge commits? Assuming you have a typical setup:

```
git rebase -i master
git push -f origin winArm64
```

**Comment by rahku** [comment]
done

File: CMakeLists.txt [hunk 1]
```diff
@@ -141,7 +144,7 @@ function(preprocess_def_file inputFilename outputFilename)
 
   add_custom_command(
     OUTPUT ${outputFilename}
-    COMMAND ${CMAKE_CXX_COMPILER} /P /EP /TC ${PREPROCESS_DEFINITIONS}  /Fi${outputFilename}  ${inputFilename}
+    COMMAND cl.exe /P /EP /TC ${PREPROCESS_DEFINITIONS}  /Fi${outputFilename}  ${inputFilename}
```
Comment (manu-silicon) [inline CMakeLists.txt]: Should `cl.exe' be really hardcoded? Same goes for other binaries that got converted to a hardcoded name with .exe.

File: build.cmd [hunk 2]
```diff
@@ -174,7 +178,11 @@ exit /b 1
 REM Build CoreCLR
 :BuildCoreCLR
 set "__CoreCLRBuildLog=%__LogsDir%\CoreCLR_%__BuildOS%__%__BuildArch%__%__BuildType%.log"
-%_msbuildexe% "%__IntermediatesDir%\install.vcxproj" %__MSBCleanBuildArgs% /nologo /maxcpucount /nodeReuse:false /p:Configuration=%__BuildType% /p:Platform=%__BuildArch% /fileloggerparameters:Verbosity=normal;LogFile="%__CoreCLRBuildLog%"
+if /i "%__BuildArch%" == "arm64" (  
+  %_msbuildexe% "%__IntermediatesDir%\install.vcxproj" %__MSBCleanBuildArgs% /nologo /maxcpucount /nodeReuse:false /p:Configuration=%__BuildType% /p:UseEnv=true /fileloggerparameters:Verbosity=normal;LogFile="%__CoreCLRBuildLog%"
+) else (
+  %_msbuildexe% "%__IntermediatesDir%\install.vcxproj" %__MSBCleanBuildArgs% /nologo /maxcpucount /nodeReuse:false /p:Configuration=%__BuildType% /p:Platform=%__BuildArch% /fileloggerparameters:Verbosity=normal;LogFile="%__CoreCLRBuildLog%"
+)
```
Comment (kangaroo) [inline build.cmd]: Why is this needed?  It seems a bit odd that BuildArch is removed?

File: src/jit/utils.h [hunk 3]
```diff
@@ -47,7 +47,7 @@ inline const char* dspBool(bool b)
 }
 
 #ifdef FEATURE_CORECLR
-#ifndef CROSSGEN_COMPILE
+//#ifndef CROSSGEN_COMPILE
```
Comment (kangaroo) [inline src/jit/utils.h]: Why is this commented and just not removed?

**Comment by kangaroo** [comment]
Has this been checked to make sure it doesn't regress the arm64/linux build?
Additionally, can't CMAKE_TOOLCHAIN_FILE be used instead of all the hardcoded names?

File: CMakeLists.txt (hunk 1 above)
Comment (janvorli) [inline CMakeLists.txt]: We should not hardcode the cl.exe. What is the problem with using the CMAKE_CXX_COMPILER?

File: CMakeLists.txt [hunk 4]
```diff
@@ -554,15 +560,23 @@ add_definitions(-DWINVER=0x0602)
 add_definitions(-DWIN32_LEAN_AND_MEAN=1)
 add_definitions(-DDEBUGGING_SUPPORTED)
 if(WIN32)
-    # Disable edit and continue on Linux
-    add_definitions(-DEnC_SUPPORTED)
+  if(CLR_CMAKE_PLATFORM_ARCH_AMD64 OR CLR_CMAKE_PLATFORM_ARCH_I386)
+      # Disable edit and continue on Linux, arm & arm64
```
Comment (janvorli) [inline CMakeLists.txt]: I know that this comment was there before, but it sounds weird. Could you please change it to "Enable edit and continue"? We don't need to enumerate the platforms, the surrounding ifs are descriptive enough.

File: src/debug/ee/wks/CMakeLists.txt [hunk 5]
```diff
@@ -15,7 +15,11 @@ add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/dbghelpers.obj
 #mark obj as source that does not require compile
 set_source_files_properties(${CMAKE_CURRENT_BINARY_DIR}/dbghelpers.obj PROPERTIES EXTERNAL_OBJECT TRUE)
```
Comment (janvorli) [inline src/debug/ee/wks/CMakeLists.txt]: It looks this should be moved next to the add_library inside the if since the obj is not used for ARM64 now.

File: src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 6]
```diff
@@ -124,22 +124,43 @@ endif(CLR_CMAKE_PLATFORM_LINUX)
...
+            COMMAND dactablegen.exe /dac:${CMAKE_CURRENT_BINARY_DIR}/daccess.i /pdb:${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/coreclr.pdb /dll:$<TARGET_FILE:coreclr> /bin:${CMAKE_CURRENT_BINARY_DIR}/wks.bin
+            COMMAND InjectResource.exe /bin:${CMAKE_CURRENT_BINARY_DIR}/wks.bin /dll:$<TARGET_FILE:coreclr>
+            COMMAND GenClrDebugResource.exe /dac:$<TARGET_FILE:mscordaccore> /dbi:$<TARGET_FILE:mscordbi> /sku:onecoreclr /out:${CMAKE_CURRENT_BINARY_DIR}/clrDebugResource.bin
+            COMMAND InjectResource.exe /bin:${CMAKE_CURRENT_BINARY_DIR}/clrDebugResource.bin /dll:$<TARGET_FILE:coreclr> /name:CLRDEBUGINFO
+            COMMENT Add dactable & debug resources to coreclr
+        )
+
+    else(CLR_CMAKE_PLATFORM_ARCH_ARM64)
+
+        add_dependencies(coreclr dactablegen)
```
Comment (janvorli) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: Why this dependency is not added for ARM64 too? The dactablegen.exe is invoked for it too, as I can see above.

File: src/inc/CMakeLists.txt [hunk 7]
```diff
@@ -30,13 +30,13 @@ get_include_directories(MIDL_INCLUDE_DIRECTORIES)
 
 
 # Run custom midl command over each idl file
-FIND_PROGRAM( MIDL midl.exe )
+#FIND_PROGRAM( MIDL midl.exe )
```
Comment (janvorli) [inline src/inc/CMakeLists.txt]: We should not hardcode the tool. Why was it needed? The midl.exe doesn't seem to be platform specific. 
Also. the FIND_PROGRAM search path is configurable if you need to let it find the exe in a different place. For example by setting the CMAKE_PROGRAM_PATH.

File: src/vm/interpreter.cpp [hunk 8]
```diff
@@ -11817,7 +11817,7 @@ void Interpreter::PrintValue(InterpreterType it, BYTE* valAddr)
                 {
                     fprintf(GetLogFile(), " ");
                 }
-                fprintf(GetLogFile(), "0x%p", valAddr[i]);
+                fprintf(GetLogFile(), "0x%b", valAddr[i]);
```
Comment (janvorli) [inline src/vm/interpreter.cpp]: It should be the original `%p`, not `%b`

**Comment by rahku** [comment]
Thanks for the review. I will try this again after I am back from my trip. closing for now. If anyone wants to use it ...can do so using winArm64 branch of my fork.

**Comment by jashook** [comment]
Reopening this pull request to start working on it some more.

I will have a change out to address comments soon.

File: src/vm/interpreter.cpp (hunk 8 above)
Comment (jashook) [inline src/vm/interpreter.cpp]: %p leads to error being treated as warning.

'fprintf'  : '%p' in format string conflicts with argument 1 of type 'int'

Comment (jkotas) [inline src/vm/interpreter.cpp]: I agree that the original `%p` is not right, but `%b` does not seem right either - it is not [documented printf string type] (https://msdn.microsoft.com/en-us/library/hf4y5e3w.aspx). I think it should be `%02x`.

Comment (jashook) [inline src/vm/interpreter.cpp]: I agree %b does not seem right either. Will change it to %02x for now.

**Comment by jashook** [comment]
Message up with the commit. Hopefully did it as expected, didn't entirely know whether to modify the original commit or no. I can change that if instructed.

@jkotas @janvorli PTAL.

File: src/dlls/mscoree/coreclr/CMakeLists.txt (hunk 6 above)
Comment (jashook) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: I do not know why this dependency was not added. However adding it causes runtime errors. I will investigate this tomorrow and hopefully have a clearer answer to why.

Comment (jashook) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: Also no, this change has not been tested with Arm64 Linux I will do that tomorrow as well.

Comment (kangaroo) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: This entire block is redundant now and should just be reverted to its original state, no?

File: src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 9]
```diff
@@ -124,20 +124,28 @@ endif(CLR_CMAKE_PLATFORM_LINUX)
...
     add_custom_command(TARGET coreclr
         POST_BUILD
         COMMAND ${CMAKE_CXX_COMPILER} /P /EP /TP ${PREPROCESS_DEFINITIONS} ${INC_DIR} /Fi${CMAKE_CURRENT_BINARY_DIR}/daccess.i  ${CLR_DIR}/src/debug/daccess/daccess.cpp
-        COMMAND $<TARGET_FILE:dactablegen_exe> /dac:${CMAKE_CURRENT_BINARY_DIR}/daccess.i /pdb:${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/coreclr.pdb /dll:$<TARGET_FILE:coreclr> /bin:${CMAKE_CURRENT_BINARY_DIR}/wks.bin
-        COMMAND $<TARGET_FILE:InjectResource> /bin:${CMAKE_CURRENT_BINARY_DIR}/wks.bin /dll:$<TARGET_FILE:coreclr>
-        COMMAND $<TARGET_FILE:GenClrDebugResource> /dac:$<TARGET_FILE:mscordaccore> /dbi:$<TARGET_FILE:mscordbi> /sku:onecoreclr /out:${CMAKE_CURRENT_BINARY_DIR}/clrDebugResource.bin
-        COMMAND $<TARGET_FILE:InjectResource> /bin:${CMAKE_CURRENT_BINARY_DIR}/clrDebugResource.bin /dll:$<TARGET_FILE:coreclr> /name:CLRDEBUGINFO
+        COMMAND dactablegen.exe /dac:${CMAKE_CURRENT_BINARY_DIR}/daccess.i /pdb:${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/coreclr.pdb /dll:$<TARGET_FILE:coreclr> /bin:${CMAKE_CURRENT_BINARY_DIR}/wks.bin
```
Comment (jashook) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: Cant seem to get rid of these hardcoded files without breaking the build for arm64 windows. Any suggestions on how to git rid of them?

File: build.cmd (hunk 2 above)
Comment (jashook) [inline build.cmd]: It is needed to get msbuild to build arm64 with the internal toolset.

Comment (jkotas) [inline build.cmd]: Add what you have just said as a comment? (Also, it should say that it should be removed once proper msbuild support is available.)

File: src/dlls/mscoree/coreclr/CMakeLists.txt (hunk 9 above)
Comment (jkotas) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: These tools would need to be packaged into a new buildtools nuget package to make this work for cross-compilation. You should comment out this section for arm64 for now and add TODO/issue for it to deal with separately.

File: CMakeLists.txt [hunk 10]
```diff
@@ -17,7 +17,7 @@ set(GENERATED_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/src/inc)
 
 function(clr_unknown_arch)
     if (WIN32)
-        message(FATAL_ERROR "Only AMD64 and I386 are supported")
+        message(FATAL_ERROR "Only AMD64, ARM64 and I386 are supported")
```
Comment (kangaroo) [inline CMakeLists.txt]: Is I386 really supported in CoreCLR currently?  Otherwise we can unify these message()'s and drop the if()

File: CMakeLists.txt [hunk 11]
```diff
@@ -375,9 +376,10 @@ set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /MANIFEST:NO") #Do n
 set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /SUBSYSTEM:WINDOWS,6.00") #windows subsystem
 set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /LARGEADDRESSAWARE") # can handle addresses larger than 2 gigabytes
 set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /RELEASE") #sets the checksum in the header
-if (IS_64BIT_BUILD EQUAL 0)
+if (CLR_CMAKE_PLATFORM_ARCH_I386 OR CLR_CMAKE_PLATFORM_ARCH_ARM64)
+set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /SAFESEH:NO")       #Image does not have Safe Exception Handlers..valid only for x86
```
Comment (kangaroo) [inline CMakeLists.txt]: Please update the comment to reflect its not just x86 anymore

File: CMakeLists.txt [hunk 12]
```diff
@@ -375,9 +376,10 @@ set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /MANIFEST:NO") #Do n
...
 set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /SAFESEH:NO") #Image does not have Safe Exception Handlers..valid only for x86
```
Comment (kangaroo) [inline CMakeLists.txt]: Ditto for the comment here please.

File: CMakeLists.txt [hunk 13]
```diff
@@ -554,15 +561,23 @@ add_definitions(-DWINVER=0x0602)
 add_definitions(-DWIN32_LEAN_AND_MEAN=1)
 add_definitions(-DDEBUGGING_SUPPORTED)
 if(WIN32)
-    # Disable edit and continue on Linux
-    add_definitions(-DEnC_SUPPORTED)
+  if(CLR_CMAKE_PLATFORM_ARCH_AMD64 OR CLR_CMAKE_PLATFORM_ARCH_I386)
+      # Enable edit and continue on Linux, arm & arm64
```
Comment (kangaroo) [inline CMakeLists.txt]: Comment doesn't seem to match the if() ?

File: src/pal/tools/gen-buildsys-win.bat [hunk 14]
```diff
@@ -25,7 +30,11 @@ if defined CMakePath goto DoGen
 for /f "delims=" %%a in ('powershell -NoProfile -ExecutionPolicy RemoteSigned "& .\probe-win.ps1"') do %%a
 
 :DoGen
-"%CMakePath%" "-DCMAKE_USER_MAKE_RULES_OVERRIDE=%basePath%\windows-compiler-override.txt" "-DCLR_CMAKE_TARGET_ARCH=%3" -G "Visual Studio %__VSString%" %1
+if "%USE_VS%" == "0" (
+    "%CMakePath%" "-DCMAKE_USER_MAKE_RULES_OVERRIDE=%basePath%\windows-compiler-override.txt" "-DCLR_CMAKE_TARGET_ARCH=%3" %1
+) else (
+    "%CMakePath%" "-DCMAKE_USER_MAKE_RULES_OVERRIDE=%basePath%\windows-compiler-override.txt" "-DCLR_CMAKE_TARGET_ARCH=%3" -G "Visual Studio %__VSString%" %1
+)
```
Comment (kangaroo) [inline src/pal/tools/gen-buildsys-win.bat]: This could probably all be cleaned up by having a environment variable for -G args, but I'm not super concerned about it.

**Comment by kangaroo** [comment]
LGTM modulo feedback, and @jkotas comments on the hard coded tools.

File: CMakeLists.txt (hunk 11 above)
Comment (jkotas) [inline CMakeLists.txt]: Does ARM64 really need the `/SAFESEH:NO` option to work? It has been legacy x86-only option since forever. I would be surprised if they brought it back for arm64.


src\dlls\mscoree\i386\handlers.asm

Comment (jkotas) [inline CMakeLists.txt]: And we should not be using /SAFESEH:NO for x86 either - opened https://github.com/dotnet/coreclr/issues/2245 on it.

File: CMakeLists.txt (hunk 10 above)
Comment (mikedn) [inline CMakeLists.txt]: It's supported in the sense that it builds successfully for x86 (or at least used to when I wrote that if).

Comment (jashook) [inline CMakeLists.txt]: Also we are focusing on x86 JIT bring up at the moment.

**Comment by jashook** [comment]
SAFESEH:NO is required workaround to a link error given by the toolset.

@jkotas @gkhanna79 PTAL

**Comment by gkhanna79** [comment]
@jashook PEr https://msdn.microsoft.com/en-us/library/9a89h429.aspx?f=255&MSPPError=-2147217396, you should not need SAFESEH as it is applicable to x86 only (as noted in the PR as well). 

Can you please share the failure you are running into?

**Comment by jashook** [comment]
@dotnet-bot test Windows_NT x86 Release

**Comment by gkhanna79** [comment]
@jashook Can you please squash the commits?

@mmitche Are the OSX build failures the issues you were looking into last week?

**Comment by mmitche** [comment]
@gkhanna79 Yeah they are.

**Comment by jashook** [comment]
@gkhanna79 Squashed

**Comment by jashook** [comment]
@jkotas @gkhanna79 PTAL. Removed SafeSEH and verified that the build is still successful.

**Comment by jashook** [comment]
@dotnet-bot test Windows x86

**Comment by gkhanna79** [comment]
The x86 Windows Debug build failed due to the following:

> 18:30:15   Read in 1611 resources from "d:\j\workspace\dotnet_coreclr\x86_debug_windows_nt_prtest\src\mscorlib\src\mscorlib.txt"
18:30:15   Writing resource file...  Done.
18:30:28   mscorlib -> d:\j\workspace\dotnet_coreclr\x86_debug_windows_nt_prtest\bin\obj\Windows_NT.x86.Debug\mscorlib.dll
18:31:03   
18:31:03   Unhandled Exception: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
18:31:03      at Microsoft.Cci.ISymUnmanagedWriter2.Close()
18:31:03      at BclRewriter.Program.RunBclRewriter(String[] args)
18:31:03      at BclRewriter.Program.Main(String[] args)


@weshaggard @AlexGhiondea Are you aware of this issue?

**Comment by gkhanna79** [comment]
@jashook Please sign the CLA. See https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing.md#contributor-license-agreement for details.

**Comment by gkhanna79** [comment]
LGTM. Please squash the commits, sign CLA and update this thread once done.

**Comment by AlexGhiondea** [comment]
@gkhanna79 no -- I have not seen that before. Does it repro consistently?

**Comment by gkhanna79** [comment]
It looks occasional - you can see the history.

**Comment by jashook** [comment]
Squashed, it also looks like Rahul needs to sign the CLA as I have already done that. Closing this PR then opening a new one to point to this.
